### PR TITLE
[3/N] Set correct device to CUDA guards

### PR DESCRIPTION
**Pull request by kwen2501**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #134707
* #134701
* __->__ #134357
* #134345

In `collective()`, `pointToPoint()` and `collectiveCoalesced()`, CUDA guards were created with an unset (default) CUDA device. This is the reason for the IMA facing the NaN checker in issue https://github.com/pytorch/pytorch/issues/134062.

With this fix, `torch.cuda.set_device(device)` is not needed to work around the IMA.

Also refactored a couple places where the guard is created -- preferably we create the guard with a known device, rather than setting the device later.

cc @XilunWu @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 1]
```diff
@@ -2132,7 +2129,9 @@ std::shared_ptr<NCCLComm> ProcessGroupNCCL::getNCCLComm(
               << timerDeltaMs << " ms";
   }
 
-  at::cuda::OptionalCUDAGuard gpuGuard;
+  // Get the device index
+  auto deviceIndex = device.index();
```
Comment (wconstab) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:2133]: Nit: do we need the deviceIndex var? Looks like we just pass device to guard directly

Ok probably we are using it somewhere, you moved the decl from below to above.

**Review by wconstab — APPROVED** [review]

**Review by wconstab — CHANGES_REQUESTED** [review]
Oh I forgot to demand tests before I stamp. Please add tests :)

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 2]
```diff
@@ -1103,11 +1103,8 @@ void ProcessGroupNCCL::abortCommsFromMap(
   for (auto& it : ncclCommsMap) {
     auto& devName = it.first;
     auto& ncclComm = it.second;
-    at::cuda::OptionalCUDAGuard gpuGuard;
     at::DeviceIndex deviceIndex = getIndexFromDeviceKey(devName);
```
Comment (shuqiangzhang) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1106]: For P2P comms, the deviceIndex could be -1 (invalid), as the keys in the map could be non deviceIndex, but rank to rank numbers.

Comment (kwen2501) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1106]: Thanks for the information. Let me revert this change here and add the above comments into the code.

**Comment by kwen2501** [comment]
> Oh I forgot to demand tests before I stamp. Please add tests :)

Testing with multiple colls now.

**Review by shuqiangzhang — APPROVED** [review]
LGTM now, thanks for the fix.

File: test/distributed/test_c10d_nccl.py [hunk 3]
```diff
@@ -388,6 +389,24 @@ def test_nan_p2p(self):
         # reset env
         os.environ["TORCH_NCCL_NAN_CHECK"] = "0"
 
+    @requires_nccl()
+    @skip_if_lt_x_gpu(2)
+    def test_nan_check(self):
+        # Not expecting an error, NaN check should not make legit code fail
+        os.environ["TORCH_NCCL_NAN_CHECK"] = "1"
```
Comment (shuqiangzhang) [inline test/distributed/test_c10d_nccl.py:396]: Should we also set CUDA_LAUNCH_BLOCKING=1 in this test? as this test failed before only it is enabled

**Review by wconstab — APPROVED** [review]

File: test/distributed/test_c10d_nccl.py (hunk 3 above)
Comment (kwen2501) [reply test/distributed/test_c10d_nccl.py:396]: I think I've seen test failing too when CUDA_LAUNCH_BLOCKING=0 (when the test has two all_reduce's).

**Pull request closed without merging**